internal/http/services/owncloud/ocdav/tpc.go

```diff
@@ -0,0 +1,418 @@
+		httpUploadRes, err := s.client.Do(httpUploadReq)
+		if err != nil {
+			return err
+		}
+
+		defer httpUploadRes.Body.Close()
+		if httpUploadRes.StatusCode != http.StatusOK {
+			return err
+		}
+	}
+	return nil
+}
+
+//
+// An example of an HTTP TPC Push
+//
+// +-----------------+        PUT          +----------------+
+// |  Dest server    |  <----------------  |  Src server    |
+// |  (Remote)       |  ---------------->  |  (Reva)        |
+// +-----------------+       Done          +----------------+
+// 												^
+// 												|
+// 												| COPY
+// 												|
+// 										   +----------+
+// 										   |  Client  |
+// 										   +----------+
+
+// handleTPCPush performs a PUT request on the remote site and while downloading
+// data from the requested reva endpoint.
+func (s *svc) handleTPCPush(ctx context.Context, w http.ResponseWriter, r *http.Request, ns string) {
+	src := path.Join(ns, r.URL.Path)
+	dst := r.Header.Get("Destination")
+	overwrite := r.Header.Get("Overwrite")
+	depth := r.Header.Get("Depth")
+	if depth == "" {
+		depth = "infinity"
+	}
+
+	sublog := appctx.GetLogger(ctx).With().Str("src", src).Str("dst", dst).Logger()
+	sublog.Debug().Str("overwrite", overwrite).Str("depth", depth).Msg("HTTPPush")
+
+	overwrite = strings.ToUpper(overwrite)
+	if overwrite == "" {
+		overwrite = "T"
+	}
+
+	if overwrite != "T" && overwrite != "F" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if depth != "infinity" && depth != "0" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// get Gateway client
+	client, err := s.getClient()
+	if err != nil {
+		sublog.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	ref := &provider.Reference{Path: src}
+	srcStatReq := &provider.StatRequest{Ref: ref}
+	srcStatRes, err := client.Stat(ctx, srcStatReq)
+	if err != nil {
+		sublog.Error().Err(err).Msg("error sending grpc stat request")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if srcStatRes.Status.Code != rpc.Code_CODE_OK && srcStatRes.Status.Code != rpc.Code_CODE_NOT_FOUND {
+		HandleErrorStatus(&sublog, w, srcStatRes.Status)
+		return
+	}
+
+	w.WriteHeader(http.StatusAccepted)
+	err = s.performHTTPPush(ctx, client, r, w, srcStatRes.Info, depth == "infinity", ns)
+	if err != nil {
+		sublog.Error().Err(err).Str("depth", depth).Msg("error descending directory")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	fmt.Fprintf(w, "success: Created")
+
+}
+
+func (s *svc) performHTTPPush(ctx context.Context, client gateway.GatewayAPIClient, r *http.Request, w http.ResponseWriter, srcInfo *provider.ResourceInfo, recurse bool, ns string) error {
+	src := path.Join(ns, r.URL.Path)
+	dst := r.Header.Get("Destination")
+
+	log := appctx.GetLogger(ctx)
+	log.Debug().Str("src", src).Str("dst", dst).Msg("Performing HTTP Push")
+
+	// get download url
+	dReq := &provider.InitiateFileDownloadRequest{
```

Comment:
Won't work for directories. Depending on whether depth is infinity or not, you'll have to recurse